src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,907 @@
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore next-line
+      isV4,
+    ).toString('hex');
+    const hashStructMessageHex = sigUtil.TypedDataUtils.hashStruct(
+      primaryType,
+      message,
+      types,
+      // @types/eth-sig-util seems to be wrong
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore next-line
+      isV4,
+    ).toString('hex');
+
+    const hdPath = await this.unlockAccountByAddress(withAccount);
+    const { success, payload }: IFrameMessageResponse = await new Promise(
+      (resolve) => {
+        this._sendMessage(
+          {
+            action: IFrameMessageAction.LedgerSignTypedData,
+            params: {
+              hdPath,
+              domainSeparatorHex,
+              hashStructMessageHex,
+            },
+          },
+          (result) => resolve(result),
+        );
+      },
+    );
+
+    if (success && isSignMessageResponse(payload)) {
+      let v = parseInt(String(payload.v), 10).toString(16);
+      if (v.length < 2) {
+        v = `0${v}`;
+      }
+      const signature = `0x${payload.r}${payload.s}${v}`;
+      // @types/eth-sig-util seems to be wrong
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore next-line
+      const addressSignedWith = sigUtil.recoverTypedSignature_v4({
+        data,
+        sig: signature,
+      });
+      if (
+        ethUtil.toChecksumAddress(addressSignedWith) !==
+        ethUtil.toChecksumAddress(withAccount)
+      ) {
+        throw new Error('Ledger: The signature doesnt match the right address');
+      }
+      return signature;
+    }
+    throw (
+      payload.error || new Error('Ledger: Unknown error while signing message')
+    );
+  }
+
+  exportAccount() {
+    throw new Error('Not supported on this device');
+  }
+
+  forgetDevice() {
+    this.accounts = [];
+    this.page = 0;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.accountDetails = {};
+    this.hdk = new HDKey();
+  }
+
+  /* PRIVATE METHODS */
+
+  #setupIframe() {
+    this.iframe = document.createElement('iframe');
+    this.iframe.src = this.bridgeUrl;
+    this.iframe.allow = `hid 'src'`;
+    this.iframe.onload = async () => {
+      // If the ledger live preference was set before the iframe is loaded,
+      // set it after the iframe has loaded
+      this.iframeLoaded = true;
+      if (this.delayedPromise) {
+        try {
+          const result = await this.updateTransportMethod(
+            this.delayedPromise.transportType,
+          );
+          this.delayedPromise.resolve(result);
+        } catch (e) {
+          this.delayedPromise.reject(e);
+        } finally {
+          delete this.delayedPromise;
+        }
+      }
+    };
+    document.head.appendChild(this.iframe);
+  }
+
+  #getOrigin() {
+    const tmp = this.bridgeUrl.split('/');
+    tmp.splice(-1, 1);
+    return tmp.join('/');
+  }
+
+  private _sendMessage(
```

Comment:
this should be changed to sharp notation too, but at this moment most of the tests rely on a stub to this method

(See also https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/174#discussion_r1127805932)

Reply:
we can spy and stub the iframe content window `postMessage` method and call the respective callback, instead of stubbing `sendMessage`, taking for granted that when whenever `postMessage` is called, a callback fn has been added to `keyring.messageCallbacks`, so we can manually call it passing the desired iframe message response:

```typescript
        sandbox.on(
          keyring?.iframe?.contentWindow as Window,
          'postMessage',
          (msg) => {
            keyring.messageCallbacks[msg.messageId]?.({
              ...msg,
              success: true,
              payload: expectedPayload,
            });
          },
        );
```